File: cyberarkhound/exporter.py
import json
from typing import Any, Dict, List
from .utils import sanitize_properties_for_bloodhound, get_logger


def export_opengraph_to_bloodhound_json(og: Any, external_edges: List[Dict[str, Any]], output_file: str, *, debug: bool = False, verbose: bool = True, log_level: str = "INFO") -> Dict[str, Any]:
    logger = get_logger(verbose)
    logger.info("Starting export to BloodHound JSON format")

    # Adjust progress logging frequency based on log level
    if log_level == "WARNING" or log_level == "ERROR":
        node_interval = 10000
        edge_interval = 50000
    elif log_level == "DEBUG":
        node_interval = 25
        edge_interval = 100
    else:  # INFO (default)
        node_interval = 100
        edge_interval = 500

    # Extract nodes
    nodes_array: List[Dict[str, Any]] = []
    if hasattr(og, 'nodes'):
        if isinstance(og.nodes, dict):
            nodes_to_process = list(og.nodes.values())
        else:
            nodes_to_process = list(og.nodes)
    else:
        nodes_to_process = []

    total_nodes = len(nodes_to_process)
    logger.info("Processing %d nodes...", total_nodes)
    
    for idx, node in enumerate(nodes_to_process, 1):
        if isinstance(node, str):
            continue
        
        # Progress logging based on log level
        if idx % node_interval == 0 or idx == total_nodes:
            logger.info("  Processed %d/%d nodes (%.1f%%)", idx, total_nodes, (idx/total_nodes)*100)
        
        node_dict = {
            "id": getattr(node, 'id', str(node)),
            "kinds": [],
            "properties": {}
        }
        kinds = getattr(node, 'kinds', [])
        if isinstance(kinds, str):
            node_dict["kinds"] = [kinds]
        else:
            node_dict["kinds"] = list(kinds)
        
        # Extract properties efficiently
        props_extracted = {}
        props = getattr(node, 'properties', None)
        if isinstance(props, dict):
            props_extracted = props
        elif hasattr(props, 'to_dict'):
            try:
                props_extracted = props.to_dict()
            except Exception:
                pass
        elif hasattr(props, '__dict__'):
            props_extracted = {k: v for k, v in vars(props).items() if not k.startswith('_')}
        
        node_dict["properties"] = sanitize_properties_for_bloodhound(props_extracted)
        nodes_array.append(node_dict)

    # Extract edges
    edges_array: List[Dict[str, Any]] = []
    if hasattr(og, 'edges'):
        if isinstance(og.edges, dict):
            edges_to_process = list(og.edges.values())
        else:
            edges_to_process = list(og.edges)
    else:
        edges_to_process = []

    total_internal_edges = len(edges_to_process)
    logger.info("Processing %d internal edges...", total_internal_edges)

    for idx, edge in enumerate(edges_to_process, 1):
        if isinstance(edge, str):
            continue
        
        # Progress logging based on log level
        if idx % edge_interval == 0 or idx == total_internal_edges:
            logger.info("  Processed %d/%d edges (%.1f%%)", idx, total_internal_edges, (idx/total_internal_edges)*100)
        
        start_node_value = getattr(edge, 'start_node', None)
        end_node_value = getattr(edge, 'end_node', None)
        if not start_node_value or not end_node_value:
            continue
        start_match_by = getattr(edge, 'start_match_by', 'id') or 'id'
        end_match_by = getattr(edge, 'end_match_by', 'id') or 'id'
        edge_dict = {
            "kind": getattr(edge, 'kind', 'Unknown'),
            "start": {"value": start_node_value, "match_by": start_match_by},
            "end": {"value": end_node_value, "match_by": end_match_by},
        }
        
        # Extract properties efficiently
        props_extracted = {}
        props = getattr(edge, 'properties', None)
        if isinstance(props, dict):
            props_extracted = props
        elif hasattr(props, 'to_dict'):
            try:
                props_extracted = props.to_dict()
            except Exception:
                pass
        elif hasattr(props, '__dict__'):
            props_extracted = {k: v for k, v in vars(props).items() if not k.startswith('_')}
        
        if props_extracted:
            edge_dict["properties"] = sanitize_properties_for_bloodhound(props_extracted)
        edges_array.append(edge_dict)

    # Add external edges
    logger.info("Processing %d external edges...", len(external_edges))
    for ext_edge in external_edges:
        edge_dict = {
            "kind": ext_edge["kind"],
            "start": ext_edge["start"],
            "end": ext_edge["end"],
        }
        if ext_edge.get("properties"):
            edge_dict["properties"] = sanitize_properties_for_bloodhound(ext_edge["properties"])
        edges_array.append(edge_dict)

    # Build final structure
    output_json = {"graph": {"nodes": nodes_array, "edges": edges_array}}
    
    # Write to file with progress indicator
    logger.info("Writing JSON to file: %s", output_file)
    logger.info("  Total nodes: %d, Total edges: %d", len(nodes_array), len(edges_array))
    
    with open(output_file, 'w', encoding='utf-8', buffering=1048576) as f:
        # Always use compact encoding for BloodHound (doesn't need pretty JSON)
        logger.info("  Writing compact JSON format...")
        json.dump(output_json, f, ensure_ascii=False, separators=(',', ':'))
    
    logger.info("Export complete! File written successfully.")
    return output_json

File: pkg/exporter/exporter.go
// Package exporter provides functions to export CyberArk OpenGraph data
// to BloodHound-compatible JSON format with progress logging.
package exporter

import (
	"bufio"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"sort"
	"strings"

	"github.com/siemens-healthineers/cyberarkhound/pkg/graph"
	"github.com/sirupsen/logrus"
)

// edgeSortKey produces a stable ordering key for an edge so that exports are
// deterministic across runs (two collections of the same environment diff cleanly).
func edgeSortKey(e *graph.Edge) string {
	propsJSON, _ := json.Marshal(e.Props)
	return strings.Join([]string{e.Kind, e.Start.Value, e.End.Value, string(propsJSON)}, "|")
}

// sortedNodes returns the graph's nodes ordered by ID for deterministic output.
func sortedNodes(og *graph.OpenGraph) []*graph.Node {
	nodes := make([]*graph.Node, 0, len(og.Nodes))
	for _, n := range og.Nodes {
		nodes = append(nodes, n)
	}
	sort.Slice(nodes, func(i, j int) bool { return nodes[i].ID < nodes[j].ID })
	return nodes
}

// sortEdgesStable sorts an edge slice in place by a precomputed stable key,
// computing each edge's key exactly once (decorate-sort) so very large graphs
// are not penalised by repeated JSON marshaling inside the comparison.
func sortEdgesStable(edges []*graph.Edge) {
	keys := make(map[*graph.Edge]string, len(edges))
	for _, e := range edges {
		keys[e] = edgeSortKey(e)
	}
	sort.SliceStable(edges, func(i, j int) bool { return keys[edges[i]] < keys[edges[j]] })
}

// buildEdgeDict serializes an edge to a BloodHound-compatible map.
//
// Edge documentation (overview, windows/linux abuse, OPSEC, references) is no
// longer copied onto every edge instance. As of BloodHound v9.5 (2026-07-29)
// that curated context lives once per relationship kind in the OpenGraph schema
// (extension/schema.json "info" sections) and BloodHound serves it from the
// entity lookup APIs with include-info=true. Keeping it out of per-edge
// properties avoids repeating the same large text blocks across every edge and
// keeps exports lean. Only the edge's own data properties are emitted here.
func buildEdgeDict(edge *graph.Edge) map[string]interface{} {
	edgeDict := map[string]interface{}{
		"kind": edge.Kind,
		"start": map[string]string{
			"value":    edge.Start.Value,
			"match_by": edge.Start.MatchBy,
		},
		"end": map[string]string{
			"value":    edge.End.Value,
			"match_by": edge.End.MatchBy,
		},
	}

	if len(edge.Props) > 0 {
		props := make(map[string]interface{}, len(edge.Props))
		for k, v := range edge.Props {
			props[k] = v
		}
		edgeDict["properties"] = props
	}

	return edgeDict
}

// ExportToBloodHoundJSON exports the OpenGraph to BloodHound JSON format
func ExportToBloodHoundJSON(og *graph.OpenGraph, outputFile string, logger *logrus.Logger, debug bool, logLevel string) error {
	logger.Info("Starting export to BloodHound JSON format")

	// Adjust progress logging frequency based on log level
	var nodeInterval, edgeInterval int
	if logLevel == "WARNING" || logLevel == "ERROR" {
		nodeInterval = 10000
		edgeInterval = 50000
	} else if logLevel == "DEBUG" {
		nodeInterval = 25
		edgeInterval = 100
	} else { // INFO (default)
		nodeInterval = 100
		edgeInterval = 500
	}

	// Sort edges for deterministic, diff-friendly output.
	sortEdgesStable(og.InternalEdges)
	sortEdgesStable(og.ExternalEdges)

	totalNodes := len(og.Nodes)
	totalInternalEdges := len(og.InternalEdges)
	totalExternalEdges := len(og.ExternalEdges)

	// Write to file.
	//
	// The graph can contain millions of nodes and edges. Materializing every
	// element as a map[string]interface{}, concatenating them into one slice
	// and handing the whole structure to json.Encoder.Encode forces the entire
	// serialized document (plus intermediate marshaling buffers) to live in
	// memory at once, which exhausts RAM on large environments. Instead we
	// stream the JSON element by element: each node/edge is marshaled,
	// written and then discarded, so peak memory stays flat regardless of
	// graph size.
	logger.Infof("Writing to file: %s", outputFile)
	file, err := os.Create(outputFile)
	if err != nil {
		return fmt.Errorf("failed to create output file: %w", err)
	}
	defer func() {
		if closeErr := file.Close(); closeErr != nil {
			logger.Warnf("Failed to close output file: %v", closeErr)
		}
	}()

	// Buffer writes so each small Marshal result doesn't translate into a
	// syscall; 1 MiB keeps throughput high without notable memory cost.
	w := bufio.NewWriterSize(file, 1<<20)

	if _, err := io.WriteString(w, `{"metadata":{"source_kind":"CyberArkBase"},"graph":{"edges":[`); err != nil {
		return fmt.Errorf("failed to write JSON: %w", err)
	}

	// writeElement marshals a single value and appends it to the current JSON
	// array, inserting a separating comma before every element after the first.
	edgeWritten := false
	writeEdge := func(edge *graph.Edge) error {
		b, err := json.Marshal(buildEdgeDict(edge))
		if err != nil {
			return err
		}
		if edgeWritten {
			if err := w.WriteByte(','); err != nil {
				return err
			}
		}
		edgeWritten = true
		_, err = w.Write(b)
		return err
	}

	// Stream internal edges.
	logger.Infof("Processing %d internal edges...", totalInternalEdges)
	for idx, edge := range og.InternalEdges {
		if (idx+1)%edgeInterval == 0 || idx+1 == totalInternalEdges {
			logger.Infof("  Processed %d/%d edges (%.1f%%)", idx+1, totalInternalEdges, float64(idx+1)/float64(totalInternalEdges)*100)
		}
		if err := writeEdge(edge); err != nil {
			return fmt.Errorf("failed to write JSON: %w", err)
		}
	}

	// Stream external edges into the same array.
	if totalExternalEdges > 0 {
		logger.Infof("Processing %d external edges...", totalExternalEdges)
		for idx, edge := range og.ExternalEdges {
			if (idx+1)%edgeInterval == 0 || idx+1 == totalExternalEdges {
				logger.Infof("  Processed %d/%d external edges (%.1f%%)", idx+1, totalExternalEdges, float64(idx+1)/float64(totalExternalEdges)*100)
			}
			if err := writeEdge(edge); err != nil {
				return fmt.Errorf("failed to write JSON: %w", err)
			}
		}
	}

	// Close the edges array and open the nodes array.
	if _, err := io.WriteString(w, `],"nodes":[`); err != nil {
		return fmt.Errorf("failed to write JSON: %w", err)
	}

	// Stream nodes (sorted by ID for deterministic output).
	logger.Infof("Processing %d nodes...", totalNodes)
	nodeWritten := false
	idx := 0
	for _, node := range sortedNodes(og) {
		idx++
		if idx%nodeInterval == 0 || idx == totalNodes {
			logger.Infof("  Processed %d/%d nodes (%.1f%%)", idx, totalNodes, float64(idx)/float64(totalNodes)*100)
		}

		nodeDict := map[string]interface{}{
			"id":         node.ID,
			"kinds":      node.Kinds,
			"properties": node.Properties,
		}
		b, err := json.Marshal(nodeDict)
		if err != nil {
			return fmt.Errorf("failed to write JSON: %w", err)
		}
		if nodeWritten {
			if err := w.WriteByte(','); err != nil {
				return fmt.Errorf("failed to write JSON: %w", err)
			}
		}
		nodeWritten = true
		if _, err := w.Write(b); err != nil {
			return fmt.Errorf("failed to write JSON: %w", err)
		}
	}

	// Close the nodes array, graph object and root object.
	if _, err := io.WriteString(w, "]}}\n"); err != nil {
		return fmt.Errorf("failed to write JSON: %w", err)
	}

	if err := w.Flush(); err != nil {
		return fmt.Errorf("failed to flush JSON: %w", err)
	}

	totalEdges := totalInternalEdges + totalExternalEdges
	logger.Infof("Export complete: nodes=%d internal_edges=%d external_edges=%d total=%d",
		totalNodes, totalInternalEdges, totalExternalEdges, totalEdges)

	return nil
}
